src/Grpc.AspNetCore.Web/Internal/GrpcWebMiddleware.cs

```diff
@@ -0,0 +1,165 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Threading.Tasks;
+using Grpc.Shared;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    internal sealed class GrpcWebMiddleware
+    {
+        private readonly GrpcWebOptions _options;
+        private readonly ILogger<GrpcWebMiddleware> _logger;
+        private readonly RequestDelegate _next;
+
+        public GrpcWebMiddleware(IOptions<GrpcWebOptions> options, ILogger<GrpcWebMiddleware> logger, RequestDelegate next)
+        {
+            _options = options.Value;
+            _logger = logger;
+            _next = next;
+        }
+
+        public Task Invoke(HttpContext httpContext)
+        {
+            var mode = GetGrpcWebMode(httpContext);
+            if (mode != GrpcWebMode.None)
+            {
+                Log.DetectedGrpcWebRequest(_logger, httpContext.Request.ContentType);
+
+                var metadata = httpContext.GetEndpoint()?.Metadata.GetMetadata<IGrpcWebEnabledMetadata>();
+                if (metadata?.GrpcWebEnabled ?? _options.GrpcWebEnabled)
+                {
+                    return HandleGrpcWebRequest(httpContext, mode);
+                }
+
+                Log.GrpcWebRequestNotProcessed(_logger);
+            }
+
+            return _next(httpContext);
+        }
+
+        private async Task HandleGrpcWebRequest(HttpContext httpContext, GrpcWebMode mode)
+        {
+            var feature = CreateAndSetGrpcWebFeature(mode, httpContext);
+
+            var initialProtocol = httpContext.Request.Protocol;
+
+            // Modifying the request is required to stop Grpc.AspNetCore.Server from rejecting it
+            httpContext.Request.Protocol = GrpcWebProtocolConstants.Http2Protocol;
+            httpContext.Request.ContentType = ResolveContentType(GrpcWebProtocolConstants.GrpcContentType, httpContext.Request.ContentType);
+
+            // Update response content type back to gRPC-Web
+            httpContext.Response.OnStarting(() =>
+            {
+                // Reset request protocol back to its original value. Not doing this causes a 2 second
+                // delay when making HTTP/1.1 calls.
+                httpContext.Request.Protocol = initialProtocol;
```

Comment:
@Tratcher @halter73 I noticed a perf issue when using this middleware with HTTP/1.1. Resetting the protocol fixed it. What was the cause?

![image](https://user-images.githubusercontent.com/303201/71223011-669d9c00-2337-11ea-98a8-80dd2e5bfc9a.png)

Reply:
It looks like it's trying to do one request per connection? I wonder if Kestrel's response logic uses the modifiable protocol field, it shouldn't.

Secondary.. Is your sample listening on IPv6? It looks like it's only listening on IPv4.

Reply:
Definitely looks like a bug: https://github.com/aspnet/AspNetCore/issues/17975

Reply:
> Secondary.. Is your sample listening on IPv6? It looks like it's only listening on IPv4.

It is setting up Kestrel endpoint with: `Listen(new IPEndPoint(IPAddress.Loopback, 5000))`. Is the two second delay when a connection starts HttpClient attempting to start a v6 connection and failing?

Reply:
> Is the two second delay when a connection starts HttpClient attempting to start a v6 connection and failing?

Yeah, you can see it in the timestamps. Always use ListenLocalhost(port) or ListenAny(port) to get both IPv4 and IPv6.